### PR TITLE
[tune] Relative logdir paths in trials for ExperimentAnalysis in remote buckets

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

When running an experiment for example in the cloud and syncing to a bucket the `logdir` path in the trials will be changed when working with the checkpoints in the bucket. There are some workarounds, but the easier solution is to also add a `rel_logdir` containing the relative path to the trials/checkpoints that can handle any changes in the location of experiment results.

As discussed with @Yard1 and @krfricke 

## Related issue number

Closes [issue 5783](https://discuss.ray.io/t/experimentanalysis-on-google-cloud-storage/5783) on the discussion board

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simonsays1980 — COMMENTED** [review]
Good one @Yard1!

File: python/ray/tune/trial.py
Comment (Yard1) [inline python/ray/tune/trial.py:218]: This has been misdocumented previously.

**Review by simonsays1980 — COMMENTED** [review]
I was unsure, if I could change `self._local_base_dir`  to a `Path` object, so left it as it was. The backward compatability was indeed something missing. Thanks for the fix, @Yard1

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:772]: Should this be
```suggestion
                else t.logdir
```
?

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:785]: Same here

```suggestion
                else checkpoint["logdir"]
```

**Review by krfricke — CHANGES_REQUESTED** [review]
Thanks for this contribution!

My main comment is that we will now have two properties with a very similar content, which seems wasteful. We will never check these for validity.

Instead, my suggestion would be the following:

1. We only store the relative logdir path (let's call it `relative_logdir`)
2. `Trial.logdir` becomes a [cached property](https://docs.python.org/3/library/functools.html#functools.cached_property) (use [this](https://pypi.org/project/cached-property/) for python 3.7 compatibility, similar to dataclasses for 3.6)
3. For backwards compatibility, we should extract the relative path when `self.logdir` is set (`@property.setter`), but print a deprecation warning
4. Same for restoring pickled objects (if `state["logdir"]` is set, just call `self.logdir = state.pop("logdir")`

We should add (very short and simple) unit tests to show that these conversions work.

It would also be great to see a test where we move an experiment analysis directory and recover the same information, showing that relative path lookup works.

Let me know if you have bandwidth to add this, otherwise I'm happy to chime in!

**Comment by Yard1** [comment]
@krfricke we don't really need to cache logdir, do we? Instead we can store both the absolute base dir path, and the relative path, and then have `logdir` join them together.

File: python/ray/tune/analysis/experiment_analysis.py
Comment (Yard1) [reply python/ray/tune/analysis/experiment_analysis.py:785]: In this case we should turn the if into `if checkpoint.get("rel_logdir", None)` for backwards compat

**Comment by simonsays1980** [comment]
> Thanks for this contribution!
> 
> My main comment is that we will now have two properties with a very similar content, which seems wasteful. We will never check these for validity.
> 
> Instead, my suggestion would be the following:
> 
> 1. We only store the relative logdir path (let's call it `relative_logdir`)
> 2. `Trial.logdir` becomes a [cached property](https://docs.python.org/3/library/functools.html#functools.cached_property) (use [this](https://pypi.org/project/cached-property/) for python 3.7 compatibility, similar to dataclasses for 3.6)
> 3. For backwards compatibility, we should extract the relative path when `self.logdir` is set (`@property.setter`), but print a deprecation warning
> 4. Same for restoring pickled objects (if `state["logdir"]` is set, just call `self.logdir = state.pop("logdir")`
> 
> We should add (very short and simple) unit tests to show that these conversions work.
> 
> It would also be great to see a test where we move an experiment analysis directory and recover the same information, showing that relative path lookup works.
> 
> Let me know if you have bandwidth to add this, otherwise I'm happy to chime in!

@krfricke I am up to finalize this PR today or tomorrow. Regarding the `cached_property` I wait for a final decision. In my opinion the backwards compatability is most important.

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:772]: Maybe just `hasattr(t, "rel_logdir")` for the if clause?

**Comment by simonsays1980** [comment]
> For @cached_property I meant exactly what @Yard1 was suggesting, but caching the results. Let's skip the caching for now though, it's just a call to `os.path.join` and shouldn't introduce a lot of overhead.
> 
> For `create_logdir`, let's change the function a bit. Pseudo code:
> 
> ```
> def init_logdir(self):
>     if not self.relative_logdir:
>         self.relative_logdir = create_unique_logdir_name(self.local_dir, self._generate_dirname())
>     assert self.logdir
>     logdir_path = Path(self.logdir)
>     logdir_path.mkdir(parents=True)
> 
>     self.invalidate_json_state()
> 
> @property
> def logdir(self) -> str:
>     if not self.relative_logdir:
>         return None
>     return os.path.join(self.local_dir, self.relative_logdir)
> 
> @logdir.setter
> def logdir(self, logdir):
>     relative_logdir = os.path.relpath(logdir, self.local_dir)
>     # assert no ".." in relative_logdir
>     logger.warning("Deprecated...") # ...
>     self.relative_logdir = relative_logdir
> 
> def create_unique_logdir_name(root: str, relative_logdir: str) -> str:
>     candidate = os.path.join(root, relative_logdir)
>     if os.path.exists(candidate):
>          # see create_logdir
>          relative_logdir = something_else
>     # Do not _create_ directory here, just return the name
>     return relative_logdir
> ```
> 
> Feel free to change the APIs as appropriate, use pathlib instead of os.path if desired, etc, this is just a suggestion

Thanks for the suggestion @krfricke. I will consider it when back at coding. Right now I am still trying to get my environment run again.

File: python/ray/tune/analysis/experiment_analysis.py
Comment (simonsays1980) [reply python/ray/tune/analysis/experiment_analysis.py]: I have used your suggestions, @krfricke

File: python/ray/tune/analysis/experiment_analysis.py
Comment (simonsays1980) [reply python/ray/tune/analysis/experiment_analysis.py]: I have used now `relative_logdir` and `logdir` in the code blocks above.

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:667]: For backwards compatibility

```suggestion
        if best_trial and hasattr(best_trial, "relative_logdir") and best_trial.relative_logdir:
```

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:774]:
```suggestion
                if hasattr(t, "relative_logdir") and t.relative_logdir
```

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:787]:
```suggestion
                if checkpoint.get("relative_logdir")
```

File: python/ray/tune/trial.py
Comment (krfricke) [inline python/ray/tune/trial.py:213]:
```suggestion
        relative_logdir: Same as ``logdir``, but relative to the parent of
```

File: python/ray/tune/trial.py
Comment (krfricke) [inline python/ray/tune/trial.py:461]: Actually I think we can remove the deprecation warning for now. I think getting the absolute logdir as an utility is quite useful

File: python/ray/tune/trial.py
Comment (krfricke) [inline python/ray/tune/trial.py:457]: Because this can potentially be set by a user/developer, let's not assert here but raise a ValueError if this is not the case

**Review by krfricke — CHANGES_REQUESTED** [review]
Few comments for backwards compatibility.

The test looks great! Yyou'll just have to add it to the `tune/BUILD` file to make sure it's executed in the CI (just put it in there in alphabetical order).

File: python/ray/tune/analysis/experiment_analysis.py
Comment (Yard1) [reply python/ray/tune/analysis/experiment_analysis.py:667]: I'd suggest for brevity:
```
if best_trial and getattr(best_trial, "relative_logdir", None):
```

File: python/ray/tune/analysis/experiment_analysis.py
Comment (Yard1) [reply python/ray/tune/analysis/experiment_analysis.py:787]: this is not going to change anything without a default value for get, right?

Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:787]: Compare

```
>>> bool({"test":1}["not_found"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'not_found'
>>> bool({"test":1}.get("not_found"))
False
```

File: python/ray/tune/analysis/experiment_analysis.py
Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:667]: Updated

File: python/ray/tune/analysis/experiment_analysis.py
Comment (Yard1) [reply python/ray/tune/analysis/experiment_analysis.py:787]: Ah, good to know!

File: python/ray/tune/trial.py
Comment (Yard1) [reply python/ray/tune/trial.py:461]: We should have this warning in the setter instead IMO

Comment (Yard1) [reply python/ray/tune/trial.py:461]: It does make sense to keep it in the setter though, right?

Comment (krfricke) [reply python/ray/tune/trial.py:461]: Yes, that makes sense! Please guard with `log_once` please

File: python/ray/tune/analysis/experiment_analysis.py
Comment (simonsays1980) [reply python/ray/tune/analysis/experiment_analysis.py:667]: Done

File: python/ray/tune/analysis/experiment_analysis.py
Comment (simonsays1980) [reply python/ray/tune/analysis/experiment_analysis.py:774]: Done

File: python/ray/tune/analysis/experiment_analysis.py
Comment (simonsays1980) [reply python/ray/tune/analysis/experiment_analysis.py:787]: Done

File: python/ray/tune/trial.py
Comment (simonsays1980) [reply python/ray/tune/trial.py:213]: Done

File: python/ray/tune/trial.py
Comment (simonsays1980) [reply python/ray/tune/trial.py:457]: Done

Comment (simonsays1980) [reply python/ray/tune/trial.py:457]: Also changed the `test_trial_relative_logdir:testDotsInLogdir` accordingly

File: python/ray/tune/trial.py
Comment (Yard1) [inline python/ray/tune/trial.py:457]:
```suggestion
        if ".." in str(relative_logdir):
```
would that work?

Comment (simonsays1980) [reply python/ray/tune/trial.py]: @Yard1 thanks for the suggestion. I included it into the last commit.

File: python/ray/tune/tests/test_trial_relative_logdir.py
Comment (krfricke) [inline python/ray/tune/tests/test_trial_relative_logdir.py:42]: `local_dir` can be undefined here

**Comment by simonsays1980** [comment]
DO we need to change `"rel_logdir"` here also into `"relative_logdir"`?

**Comment by Yard1** [comment]
@simonsays1980 yes, it has to be changed

File: python/ray/tune/trial.py
Comment (krfricke) [inline python/ray/tune/trial.py:458]: Let's raise a more actionable error message here:

```suggestion
            raise ValueError(
                f"The `logdir` points to a directory outside the trial's `local_dir` "
                f"({self.local_dir}), which is unsupported. Use a logdir within the "
                f"local directory instead. Got: {logdir}"
            )
```

**Review by krfricke — COMMENTED** [review]
Almost there, a few nits, and tests are currently failing:

TrialRelativeLogdirTest::testRelativeLogdir
AnalysisSuite::testBestConfigIsLogdir
AnalysisSuite::testDataframe
ExperimentAnalysisSuite::testTrialDataframe

File: python/ray/tune/tests/test_trial_relative_logdir.py
Comment (krfricke) [inline python/ray/tune/tests/test_trial_relative_logdir.py:63]: `local_dir` can be undefined here

File: python/ray/tune/trial.py
Comment (simonsays1980) [reply python/ray/tune/trial.py:459]: Done

File: python/ray/tune/trial.py
Comment (simonsays1980) [reply python/ray/tune/trial.py:458]: Done

**Comment by krfricke** [comment]
We're righting with a protobuf version upgrade at the moment, can you merge latest master so CI kicks off again? Thanks!

**Comment by simonsays1980** [comment]
> We're righting with a protobuf version upgrade at the moment, can you merge latest master so CI kicks off again? Thanks!

I am fighting with somehow some mess due to git I guess. Changes from before (e.g. the ones from @Yard1 were lost and now and I coded them from anew. I see that the tests that failed, fail due to the fact that the `local_base_dir` in the `ExperimentAnalysis` keeps the same even though there are two experiments. That fails then in `get_all_configs()` to find the configs for the other experiment as they are in the first folder (`analysis_exp1`). 

I can definitely merge latest master, but what I push then will definitely fail the tests.

**Comment by simonsays1980** [comment]
> > We're righting with a protobuf version upgrade at the moment, can you merge latest master so CI kicks off again? Thanks!
> 
> I am fighting with somehow some mess due to git I guess. Changes from before (e.g. the ones from @Yard1 were lost and now and I coded them from anew. I see that the tests that failed, fail due to the fact that the `local_base_dir` in the `ExperimentAnalysis` keeps the same even though there are two experiments. That fails then in `get_all_configs()` to find the configs for the other experiment as they are in the first folder (`analysis_exp1`).
> 
> I can definitely merge latest master, but what I push then will definitely fail the tests.

I guess the problem we are facing is that we are missing the experiment name directory in the `_local_base_dir`-attribute of the `ExperimentAnalysis`. What happens is: the `Trial` contains it still as the leaf directory in its `local_dir`-attribute. Given this `local_dir` we define the `relative_logdir` without any experiment name (or for PPO the "PPO" directory). Then in the `ExperimentAnalysis` we build the `trial_path`s with the `ExperimentAnalysis`'s `_local_base_dir` and the `Trial`'s `relative_logdir` - as a result the expeirment directory is missing in between and the tests fail. 

I try to find a fix. Any suggestions welcome :)

**Comment by simonsays1980** [comment]
Okay, this is a good test as it shows that the approach so far misses consistency :) 
I isolated the error a little that occurrs in a specific situation created by `testBestConfigIsLogdir`: 

## What causes the error
In this situation we have two experiments (`analysis_exp1` and `analysis_exp2`) in the same base directory. The `experiment_checkpoint_path` passed to the `ExperimentAnalysis` constructor is however the base directory itself (e.g. `/tmp/tmp5tszah`). We define then the `_local_base_dir` as `Path(latest_checkpoint[0]).parent` (this is crucial for the error!). 
Then the `_get_trial_paths()` (for `get_all_configs()`) and `get_best_logdir()` get the trial paths from joining `self._base_local_dir` (based on the latest checkpoint's path) with `trial.relative_logdir`. 
If now the best trial **by case** is stored in the same experiment folder (e.g. `analysis_exp2`) as the latest checkpoint everything runs fine - if not: we get the error. 

## What are the options?
I see here two main options: 

1. We can use a similar pattern as the one that collects the `_experiment_states` (as is is looked for the `.json` files also in subdirs and where these files are the trials must be). This happens in `get_latest_checkpoint()` where checkpoints get collected. We then should also change the `self._local_base_dir` to `Path(latest_checkpoint[0]).parents[1]` to get the root dir instead of the experiment folder (setting it now to this path will result in the experiment folder removed completely from the path which will result in an error, e.g. `get_all_configs()` returns an empty dictionary).
2. We pass to each trial another attribute which is the experiment name (the `name` passed to `tune.run()`) and use this together with the `self._local_base_dir` and `trial.relative_logdir`. This attribute has to be set at a point where in `tune.run()` trials are accessible (not necessary created as this happens in the search algorithms as far as I understand). 

No matter which option we choose: we have to prescribe a certain pattern how trials/experiments can be moved such that reading the moved trials/experiment still works. These rules might be something like: 

- *"If you set a name for your experiment you need to move this folder together with the trials."*

Open to discussion

**Comment by krfricke** [comment]
IMO we shouldn't go for 2, as the `Trial` itself should not have to know about the concept of Experiments. 

Actually I think the solution could be quite easy. I think we don't actually need this:

```
            _trial_paths = [
                str(self._local_base_dir.joinpath(t.relative_logdir))
                if getattr(t, "relative_logdir", None)
                else str(t.logdir)
                for t in self.trials
            ]
```

and instead can use

```
            _trial_paths = [
                str(t.logdir)
                for t in self.trials
            ]
```

as the `t.logdir` will be resolved on the `Trial` object side.

The only thing we have to do is when we parse a directory and re-constuct Trial stubs to update the trial local_dir. Then we should be good, right?

**Comment by Yard1** [comment]
Alternatively, would something like this work (assuming `local_base_dir` points to the root dir and not experiment dir)?
```python
            _trial_paths = [
                str(self._local_base_dir.joinpath(Path(t.local_dir).name).joinpath(t.relative_logdir))
                if getattr(t, "relative_logdir", None)
                else str(t.logdir)
                for t in self.trials
            ]
```
We know that the last component of the `local_dir` path (`.name`) will be the experiment dir. That way we don't have to modify the stubs. We can put this logic into a function and reuse it.

**Comment by simonsays1980** [comment]
> IMO we shouldn't go for 2, as the `Trial` itself should not have to know about the concept of Experiments.
> 
> Actually I think the solution could be quite easy. I think we don't actually need this:
> 
> ```
>             _trial_paths = [
>                 str(self._local_base_dir.joinpath(t.relative_logdir))
>                 if getattr(t, "relative_logdir", None)
>                 else str(t.logdir)
>                 for t in self.trials
>             ]
> ```
> 
> and instead can use
> 
> ```
>             _trial_paths = [
>                 str(t.logdir)
>                 for t in self.trials
>             ]
> ```
> 
> as the `t.logdir` will be resolved on the `Trial` object side.
> 
> The only thing we have to do is when we parse a directory and re-constuct Trial stubs to update the trial local_dir. Then we should be good, right?

@krfricke I agree on keeping the independence between `Trial` and `Experiment` classes. Taking the `local_dir` will work as long as we do not move trials/folders with trials. So we have to somehow account for the latter case. 

*Solution* might be to check the `self._local_base_dir` or `_experiment_states`' paths against the `trial.local_dir` during initialization.

Another one would be to make thorough search on the `experiment_checkpoint_path` to collect the trial paths by brute force.

**Comment by simonsays1980** [comment]
> Alternatively, would something like this work (assuming `local_base_dir` points to the root dir and not experiment dir)?
> 
> ```python
>             _trial_paths = [
>                 str(self._local_base_dir.joinpath(Path(t.local_dir).name).joinpath(t.relative_logdir))
>                 if getattr(t, "relative_logdir", None)
>                 else str(t.logdir)
>                 for t in self.trials
>             ]
> ```
> 
> We know that the last component of the `local_dir` path (`.name`) will be the experiment dir. That way we don't have to modify the stubs. We can put this logic into a function and reuse it.

Thanks for the suggestion @Yard1, in my opinion the `self._local_base_dir` should always point to the root precisely because of the cases where multiple experiments are located in the same root folder. 

I like the idea to create a function that contain the trial paths including the experiment folders. I tried out what happens when I give the experiment name `analysis_exp1/analysis_exp2` and this nesting of experiments works - it would however bring another difficulty into our search for trial paths: trials will then contain `analysis_exp1/analysis_exp3` in the `local_dir` - using `Path(..).name` on them, however, looses the first directory `analysis_exp1` leading to an error. 

- We cannot compare with the `_local_base_dir` as this changes when we move around the trials or experiments.  
- We cannot use the experiment name inside of the `Trial`s as this would then make the `Trial` class know about the former one. 
- [ ] We can however use brute force (lets call the approach *search*) to work along the given `experiment_checkpoint_path` to look for trials. 
- [ ] We could avoid costly search with more information, namely somehow getting the experiment name (here `analysis_exp1/analysis_exp2`) in the `ExperimentAnalysis` class (this class can know about experiments I guess and the trials do not need to). Is there maybe some info in the `_experiment_states` that could be (mis)used for it?

**Comment by Yard1** [comment]
I see. Yeah, you are correct, we cannot assume that the experiment dir will not be nested. I think going with Kai's suggestion would work. We already have logic inside ExperimentAnalysis to find the experiment_state.json file. We probably could use that to get the correct dirs for the trials.

**Comment by simonsays1980** [comment]
> I see. Yeah, you are correct, we cannot assume that the experiment dir will not be nested. I think going with Kai's suggestion would work. We already have logic inside ExperimentAnalysis to find the experiment_state.json file. We probably could use that to get the correct dirs for the trials.

@Yard1 The solution @krfricke suggested works in case of no movement of directories. For the other case we have to look for a solution then. 

- [ ] Checking one path (`trial.local_dir` or `trial.relative_logdir`) against another might give a hint on how the actual directory looks like. I guess looking for the `trial.relative_logdir` in the actual path could help here. 
- [ ] Checking via *search* like for the `experiment_state` files would be another. 

I think that taking on the user's perspective here could lead us to a clear solution: there are two situations here: 
1. The user gives a `experiment_checkpoint_path` pointing directly to an `experiment_state` file (and in this directory there are one or more trials) or
     - *Using then only the trials of the experiment makes sense*
2. He/she gives us an `experiment_checkpoint_path` that points to a directory containing one or more experiments/trials. 
     - *Using all trials from all subdirectories makes sense here*

I try to make some experiments with what works well ...

**Comment by Yard1** [comment]
I've posted a proposal of how we could deal with this here - https://github.com/simonsays1980/ray/pull/53

**Comment by Yard1** [comment]
Hey @simonsays1980, something seems to have gone wrong with the merge and many other files are overwritten. Can you check?

**Review by krfricke — APPROVED** [review]
This all looks great. Thank you so much for your contribution and all the iterations @simonsays1980! I've fixed the last test (only added a `trial_id` filed) and merged latest master, I'll merge once CI passes!

**Review by Yard1 — APPROVED** [review]
Looks great, thank you for iterating on this @simonsays1980 !

**Comment by simonsays1980** [comment]
@krfricke @Yard1 I am happy to have contributed. I am also looking forward to use this feature in ray tune. I think this helps a lot of users to overcome the problems I had with analysing results. 
Thanks for an excellent guidance and a lot I could learn about how you guys want to handle things in tune. All these iterations were worth it imo. We had some very good discussions where a good solution evolved from. Thanks for your patience and to @Yard1 thank you for the extra miles!

File: python/ray/tune/analysis/experiment_analysis.py
Comment (Yard1) [inline python/ray/tune/analysis/experiment_analysis.py:182]:
```suggestion
                    f"The directory `{experiment_checkpoint_path}` does not "
```

**Comment by Nintorac** [comment]
Nice! Great addition thanks :D

**Comment by simonsays1980** [comment]
I am happy for this solution! Great work @krfricke, @Yard1 !

**Pull request merged**
